### ISSUE TITLE
TensorFlow Implementation

### ISSUE THREAD

**Issue by inesani** [wuyifan18/DeepLog#29]
First of all, I want to thank @wuyifan18 for this awesome code. I am trying to translate this Pytorch code into Tensorflow and I am having some trouble. I am wondering if anyone here has done it or can take a look at my implementation. 

I have reproduced the Model class on something that I think is the equivalent:

```
window_size = 10
input_size = 1
hidden_size = 64
num_layers = 2
num_classes = 28
num_epochs = 20
num_candidates = 9
batch_size = 2048
model_path = 'model/'
log = 'TFAdam_batch_size=' + str(batch_size) + ';epoch=' + str(num_epochs)


modeltf = keras.models.Sequential([
  keras.layers.LSTM(hidden_size, return_sequences=True,
                         batch_input_shape=(batch_size, 10, 1)),
  keras.layers.LSTM(hidden_size, return_sequences=False),
  keras.layers.Dense(num_classes),
])

modeltf.compile(optimizer=keras.optimizers.Adam(),
              loss='sparse_categorical_crossentropy',
              metrics=['sparse_categorical_accuracy'])

#slight modification on generate function, for keeping np arrays instead of pytorch tensors
def generate(name):
    num_sessions = 0
    inputs = []
    outputs = []
    with open('data/' + name, 'r') as f:
        for line in f.readlines():
            num_sessions += 1
            line = tuple(map(lambda n: n - 1, map(int, line.strip().split())))
            for i in range(len(line) - window_size):
                inputs.append(line[i:i + window_size])
                outputs.append(line[i + window_size])
    print('Number of sessions({}): {}'.format(name, num_sessions))
    print('Number of seqs({}): {}'.format(name, len(inputs)))
    #dataset = [tf.convert_to_tensor(inputs), tf.convert_to_tensor(outputs)]
    dataset = [np.array(inputs), np.array(outputs)]
    return dataset

seq_dataset = generate('hdfs_train')
trainX = seq_dataset[0]
trainY = seq_dataset[1]

#I am letting down the last sessions as tf only accepts a multiple of the batch size as input size
trainX = trainX[:45056]
trainY = trainY[:45056]

trainX = np.reshape(trainX, (trainX.shape[0], trainX.shape[1], 1))

modeltf.fit(trainX, trainY, epochs=num_epochs, batch_size=batch_size)
```

The problem is that my results are not good, they do not resemble at all what the Pytorch code achieves. 
This can be seen directly from the loss and the accuracy the model displays when training. It gets stuck at around 3. I have tried batch normalization without much success and it is really bugging me because in pytorch it works perfectly. I think that it might be something subtle but I am not able to find it. 

If anyone here has done it or want to take time to take a look I will be very grateful :)

**Comment by wuyifan18** [comment]
Hi @inesani , I am not familiar with Tensorflow, I just glance at these codes but can not find any problems. I am sorry. But can you show me the loss and the accuracy the model displays when training?

**Comment by inesani** [comment]
Yeah sure. When I execute the above I get:

```
Number of sessions(hdfs_train): 4855
Number of seqs(hdfs_train): 46575

Train on 45056 samples
Epoch 1/20
45056/45056 [==============================] - 5s 115us/sample - loss: 7.6743 - sparse_categorical_accuracy: 0.1109
Epoch 2/20
45056/45056 [==============================] - 5s 112us/sample - loss: 7.4262 - sparse_categorical_accuracy: 0.0912
Epoch 3/20
45056/45056 [==============================] - 5s 101us/sample - loss: 3.9603 - sparse_categorical_accuracy: 0.1085
Epoch 4/20
45056/45056 [==============================] - 5s 100us/sample - loss: 3.3323 - sparse_categorical_accuracy: 0.0327
Epoch 5/20
45056/45056 [==============================] - 5s 102us/sample - loss: 3.3322 - sparse_categorical_accuracy: 0.0327
Epoch 6/20
45056/45056 [==============================] - 5s 104us/sample - loss: 3.3322 - sparse_categorical_accuracy: 0.0327
Epoch 7/20
45056/45056 [==============================] - 5s 105us/sample - loss: 3.3322 - sparse_categorical_accuracy: 0.0327
Epoch 8/20
45056/45056 [==============================] - 5s 107us/sample - loss: 3.3322 - sparse_categorical_accuracy: 0.0327
Epoch 9/20
45056/45056 [==============================] - 5s 106us/sample - loss: 3.3322 - sparse_categorical_accuracy: 0.0327
```

I don't print the rest because the loss gets into this 3.3322 value for ever (even with 300 epochs) and I can't figure out why the output is so different when the code 'seems right to me'.

**Comment by TiberiuCocias** [comment]
Hi all,

hope that it helps: for the layer Dense add a ‘softmax‘ activation in order to predict the probability for each class

e.g.:  keras.layers.Dense(num_classes, activation='softmax')

Using this activation i obtained: 
43008/45056 [===========================>..] - ETA: 0s - loss: 0.3886 - acc: 0.8568

**Comment by wuyifan18** [comment]
@TiberiuCocias thanks for your help~
@inesani you can refer to his answer.

**Comment by inesani** [comment]
Thank you very much @TiberiuCocias :) this simple trick has avoided the cost function to get stucked and now it decreases!! I have been reading that Pytorch applies Softmax automatically when using the CrossEntropyLoss. Thank you!!

Now the mistery remains because after happily training my new model I am making predictions on the provided test_normal and test_abnormal files and the results are far from the ones obtained with pytorch. When executing the modified code with the predictions I am getting:

false positive (FP): 5078, false negative (FN): 142774, Precision: 79.689%, Recall: 12.245%, F1-measure: 21.229%

The full code:

```
import tensorflow as tf
import numpy as np
import time
keras = tf.keras
from keras.utils import to_categorical

import os
import datetime
now = datetime.datetime.now()
print ("Start date and time : ")
print (now.strftime("%Y-%m-%d %H:%M:%S"))


window_size = 10
input_size = 1
hidden_size = 64
num_layers = 2
num_classes = 28
num_epochs = 20
num_candidates = 9
batch_size = 2048


modeltf = keras.models.Sequential([
  keras.layers.LSTM(hidden_size, return_sequences=True,
                         batch_input_shape=(batch_size, 10, 1)),
  keras.layers.LSTM(hidden_size, return_sequences=False),
  keras.layers.Dense(num_classes, activation='softmax'),
])

modeltf.compile(optimizer=keras.optimizers.Adam(),
              loss='sparse_categorical_crossentropy',
              metrics=['sparse_categorical_accuracy'])

#slight modification on generate function, for keeping np arrays instead of pytorch tensors
def generate(name):
    num_sessions = 0
    inputs = []
    outputs = []
    with open('data/' + name, 'r') as f:
        for line in f.readlines():
            num_sessions += 1
            line = tuple(map(lambda n: n - 1, map(int, line.strip().split())))
            for i in range(len(line) - window_size):
                inputs.append(line[i:i + window_size])
                outputs.append(line[i + window_size])
    print('Number of sessions({}): {}'.format(name, num_sessions))
    print('Number of seqs({}): {}'.format(name, len(inputs)))
    #dataset = [tf.convert_to_tensor(inputs), tf.convert_to_tensor(outputs)]
    dataset = [np.array(inputs, dtype=np.float32), np.array(outputs, dtype=np.float32)]
    return dataset

seq_dataset = generate('train')
trainX = seq_dataset[0]
trainY = seq_dataset[1]

#I am letting down the last sessions as tf only accepts a multiple of the batch size as input size
trainX = trainX[:45056]
trainY = trainY[:45056]

trainX = np.reshape(trainX, (trainX.shape[0], trainX.shape[1], 1))

modeltf.fit(trainX, trainY, epochs=num_epochs, batch_size=batch_size)


test_normal_loader = generate('test_normal')

normalX = test_normal_loader[0]
normalY = test_normal_loader[1]

normalX = np.reshape(normalX, (normalX.shape[0], normalX.shape[1], 1))
normalX = normalX[:5257216]
normalY = normalY[:5257216]

testnormaldataset = tf.data.Dataset.from_tensor_slices((normalX))
testnormaldataset = testnormaldataset.batch(batch_size)


predictions = modeltf.predict(testnormaldataset)

FP = 0

for p in range(predictions.shape[0]):
    predicted = np.argpartition(predictions[p], -num_candidates)[-num_candidates:]
    label = normalY[p]
    if label not in predicted:
        FP += 1


test_abnormal_loader = generate('test_abnormal')

abnormalX = test_abnormal_loader[0]
abnormalY = test_abnormal_loader[1]

abnormalX = np.reshape(abnormalX, (abnormalX.shape[0], abnormalX.shape[1], 1))
abnormalX = abnormalX[:161792]
abnormalY = abnormalY[:161792]

testabnormaldataset = tf.data.Dataset.from_tensor_slices((abnormalX))
testabnormaldataset = testabnormaldataset.batch(batch_size)

abpredictions = modeltf.predict(testabnormaldataset)

TP = 0
for p in range(abpredictions.shape[0]):
    predicted = np.argpartition(abpredictions[p], -num_candidates)[-num_candidates:]
    label = abnormalY[p]
    if label not in predicted:
        TP += 1



# Compute precision, recall and F1-measure
FN = len(test_abnormal_loader[0]) - TP
P = 100 * TP / (TP + FP)
R = 100 * TP / (TP + FN)
F1 = 2 * P * R / (P + R)
print('false positive (FP): {}, false negative (FN): {}, Precision: {:.3f}%, Recall: {:.3f}%, F1-measure: {:.3f}%'.format(
        FP, FN, P, R, F1))
print('Finished Predicting')
```

I guess that the training might be overfitting and so it generalizes badly on the testing but I can't see why on Pytorch it does give optimal results. I have been unable to correct this and I will gladly accept ideas in order to finish up the TF implementation.

**Comment by saucer-man** [comment]
> num_layers

Did you find the reason,  I also encountered a similar problem, thanks for any help

**Comment by inesani** [comment]
Unfortunately I didn't...

**Comment by saucer-man** [comment]
I took a closer look. During the detection, for each task flow, as long as a sequence is abnormal, the task flow is determined to be abnormal. Therefore, each task stream needs to be detected separately. In the version of @ wuyifan18, the two generate functions are different. Below is my simple d code

```
import numpy as np
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Dropout
from keras.layers import LSTM
from keras.callbacks import ModelCheckpoint
import time

def generate(name, window_size):
    num_sessions = 0
    inputs = []
    outputs = []
    with open('data/' + name, 'r') as f:
        for line in f.readlines():
            num_sessions += 1
            line = tuple(map(lambda n: n - 1, map(int, line.strip().split())))
            for i in range(len(line) - window_size):
                inputs.append(line[i:i + window_size])
                outputs.append(line[i + window_size])
    print('Number of sessions({}): {}'.format(name, num_sessions))
    print('Number of seqs({}): {}'.format(name, len(inputs)))
    return inputs, outputs


def lstm_train(x, y, num_epochs, batch_size):
    model = Sequential()
    model.add(LSTM(64, activation='relu', return_sequences=True, input_shape=(x.shape[1], x.shape[2])))
    model.add(LSTM(64, return_sequences=False))  
    model.add(Dense(y.shape[1], activation='softmax'))
    model.compile(loss="categorical_crossentropy", optimizer='adam', metrics=['accuracy'])
    model.fit(x, y, epochs=num_epochs, batch_size=batch_size, shuffle=True) 
    return model


if __name__ == "__main__":
    start_time = time.time()
    num_classes = 28 
    num_epochs = 50  
    batch_size = 2048  
    window_size = 10  

    TP = 0  
    FP = 0  
    n_candidates = 9  # top n probability of the next tag


    X, Y = generate('hdfs_train', window_size)
    X = np.reshape(X, (len(X), window_size, 1)) 
    X = X / float(num_classes)
    Y = np_utils.to_categorical(Y, num_classes)  
    model = lstm_train(X, Y, num_epochs, batch_size)


    def generate1(name, window_size):
        hdfs = set()
        with open('data/' + name, 'r') as f:
            for ln in f.readlines():
                ln = list(map(lambda n: n - 1, map(int, ln.strip().split())))
                ln = ln + [-1] * (window_size + 1 - len(ln))
                hdfs.add(tuple(ln))
        print('Number of sessions({}): {}'.format(name, len(hdfs)))
        return hdfs

    test_normal_loader = generate1('hdfs_test_normal', window_size)
    test_abnormal_loader = generate1('hdfs_test_abnormal', window_size)

    for line in test_abnormal_loader:
        for i in range(len(line) - window_size):
            seq = line[i: i + window_size]
            label = line[i + window_size]
            X = np.reshape(seq, (1, window_size, 1)) 
            X = X / float(num_classes)
            Y = np_utils.to_categorical(label, num_classes)  
            prediction = model.predict(X, verbose=0) 
            if np.argmax(Y) not in prediction.argsort()[0][::-1][: n_candidates]:

                TP += 1
                break

    for line in test_normal_loader:
        for i in range(len(line) - window_size):
            seq = line[i:i + window_size]
            label = line[i + window_size]
            X = np.reshape(seq, (1, window_size, 1)) 
            X = X / float(num_classes)
            Y = np_utils.to_categorical(label, num_classes) 
            prediction = model.predict(X, verbose=0)
            if np.argmax(Y) not in prediction.argsort()[0][::-1][: n_candidates]:
                FP += 1
                break

    elapsed_time = time.time() - start_time
    print('elapsed_time: {:.3f}s'.format(elapsed_time))


    # Compute precision, recall and F1-measure
    FN = len(test_abnormal_loader) - TP
    TN = len(test_normal_loader) - FP
    P = 100 * TP / (TP + FP)
    R = 100 * TP / (TP + FN)
    F1 = 2 * P * R / (P + R)

    print(f"FP:{FP}")
    print(f"FN: {FN}")
    print(f"TP: {TP}")
    print(f"TN: {TN}")

    print('false positive (FP): {}, false negative (FN): {}, Precision: {:.3f}%, Recall: {:.3f}%, F1-measure: {:.3f}%'.format(FP, FN, P, R, F1))
    print('Finished Predicting')
```
the result is false positive (FP): 908, false negative (FN): 144, Precision: 81.420%, Recall: 96.507%, F1-measure: 88.324%
